### ISSUE TITLE
Use of memory after it is freed

### ISSUE THREAD

**Issue by mself** [quicky2000/tartini#45]
The Xcode static analyzer identified a possible issue with using freed memory in `SmartPtr`:

```
tartini/widgets/vibrato/vibratowidget.cpp:272:13: Use of memory after it is freed (within a call to 'get_maxima')
In file included from tartini/widgets/vibrato/vibratowidget.cpp:21:
In file included from tartini/sound/channel.h:28:
In file included from sound/notedata.h:22:
tartini/include/SmartPtr.h:327:5: warning: Use of memory after it is freed
    return m_ptr;
    ^~~~~~~~~~~~

tartini/widgets/vibrato/vibratospeedwidget.cpp:524:42: Use of memory after it is freed (within a call to 'get_maxima')
In file included from tartini/widgets/vibrato/vibratospeedwidget.cpp:19:
In file included from tartini/sound/channel.h:28:
In file included from sound/notedata.h:22:
tartini/include/SmartPtr.h:327:5: warning: Use of memory after it is freed
    return m_ptr;
    ^~~~~~~~~~~~

tartini/widgets/vibrato/vibratoperiodwidget.cpp:173:26: Use of memory after it is freed (within a call to 'get_minima')
In file included from tartini/widgets/vibrato/vibratoperiodwidget.cpp:19:
In file included from tartini/sound/channel.h:28:
In file included from sound/notedata.h:22:
tartini/include/SmartPtr.h:327:5: warning: Use of memory after it is freed
    return m_ptr;
    ^~~~~~~~~~~~
```

**Comment by quicky2000** [comment]
Yes I saw in the report you attached to an other issue.
Are there more details?
In the past bugs reported by Clang static analyzer I encountered where provided with a kind of scenario indicating conditions to follow to make the bug occur. It was very usefull to understand and solve the bug

**Comment by mself** [comment]
Here's a screen shot showing what the static analyzer reported.  I didn't find it easy to understand.

Note that I also updated the comment above with some additional log lines that I missed the first time.

![image](https://user-images.githubusercontent.com/6620480/80769895-7d34da80-8b03-11ea-942d-792c9bf80f17.png)

**Comment by mself** [comment]
I think that the issue might be related to line 208, where `l_note` is set to a pointer to an element inside a `large_vector`:
```
	  const NoteData * l_note = &(l_active->get_note_data()[l_data->getNoteIndex()]);
```
That seems like it could be dangerous.  Maybe the element that `l_note` points to gets freed somewhere between lines 216 and 272 (although I don't see how)?

But, if so, why is there no warning when `l_note` is used again on line 310?

Maybe the static analyzer is just confused and these warnings are not valid?

**Comment by quicky2000** [comment]
the screenshot is for which file ?

**Comment by mself** [comment]
The first one: `tartini/widgets/vibrato/vibratowidget.cpp`

**Comment by mself** [comment]
I figured out how to get more detailed logs from the static analyzer, so here is some more info:

```
tartini/widgets/vibrato/vibratowidget.cpp:272:13: Use of memory after it is freed (within a call to 'get_maxima')
tartini/widgets/vibrato/vibratowidget.cpp:202:7: Assuming 'l_active' is non-null
tartini/widgets/vibrato/vibratowidget.cpp:206:10: Assuming 'l_data' is non-null
tartini/widgets/vibrato/vibratowidget.cpp:206:20: Assuming the condition is true
tartini/widgets/vibrato/vibratowidget.cpp:206:71: Assuming the condition is true
tartini/widgets/vibrato/vibratowidget.cpp:214:30: Calling 'NoteData::get_maxima'
/sound/notedata.hpp:92:1: Entered call from 'VibratoWidget::doUpdate'
/sound/notedata.hpp:94:12: Calling copy constructor for 'SmartPtr<Array1d<int>>'
tartini/include/SmartPtr.h:194:1: Entered call from 'NoteData::get_maxima'
tartini/include/SmartPtr.h:198:8: Assuming field 'm_ptr' is non-null
/sound/notedata.hpp:94:12: Returning from copy constructor for 'SmartPtr<Array1d<int>>'
tartini/widgets/vibrato/vibratowidget.cpp:214:30: Returning from 'NoteData::get_maxima'
tartini/widgets/vibrato/vibratowidget.cpp:214:49: Calling '~SmartPtr'
tartini/include/SmartPtr.h:210:1: Entered call from 'VibratoWidget::doUpdate'
tartini/include/SmartPtr.h:214:12: Assuming the condition is true
tartini/include/SmartPtr.h:217:13: Memory is released
tartini/widgets/vibrato/vibratowidget.cpp:214:49: Returning from '~SmartPtr'
tartini/widgets/vibrato/vibratowidget.cpp:224:8: Assuming the condition is false
tartini/widgets/vibrato/vibratowidget.cpp:258:9: Assuming the condition is true
tartini/widgets/vibrato/vibratowidget.cpp:258:45: Assuming the condition is true
tartini/widgets/vibrato/vibratowidget.cpp:267:12: Assuming the condition is true
tartini/widgets/vibrato/vibratowidget.cpp:270:9: Assuming 'l_maxima_size' is equal to 'l_minima_size'
tartini/widgets/vibrato/vibratowidget.cpp:272:13: Calling 'NoteData::get_maxima'
/sound/notedata.hpp:92:1: Entered call from 'VibratoWidget::doUpdate'
/sound/notedata.hpp:94:12: Calling copy constructor for 'SmartPtr<Array1d<int>>'
tartini/include/SmartPtr.h:194:1: Entered call from 'NoteData::get_maxima'
tartini/include/SmartPtr.h:195:9: Calling 'SmartPtr::ptr'
tartini/include/SmartPtr.h:325:1: Entered call from copy constructor for 'SmartPtr<Array1d<int>>'
tartini/include/SmartPtr.h:327:5: Use of memory after it is freed
```

**Comment by mself** [comment]
Maybe `NoteData::get_maxima()` should return a reference rather than making a copy of `m_maxima`?

**Comment by mself** [comment]
I think the static analyzer is just getting confused by the use of `SmartPtr` and this is not a bug.

In a call like this:
```
	  const int l_maxima_size = l_note->get_maxima()->size();
```
`get_maxima()` returns a copy of the `SmartPtr` that points to `m_maxima` inside `l_note`.  When the `SmartPtr` is copied, the reference count goes up by one.  After the call to `size()` finishes, the reference count goes down by one (when the copied `SmartPtr` goes out of scope).

The static analyzer sees where the reference count is decremented, but it doesn't know that `l_note` still holds a reference to `m_maxima` and that the underlying object will never be freed here.

The static analyzer assumes that the reference count might have gone to zero and that the object might have been deleted.  That is why it is only the *second* call to `get_maxima()` that triggers the warning.

I think we can safely close this bug, although I wish there was a way to prevent the static analyzer from flagging these cases in the future.

**Comment by mself** [comment]
I figured out a way to suppress these warnings from the static analyzer.  I'll submit a PR.

**Comment by quicky2000** [comment]
Closed by Pull Request #47
